stdlib/public/runtime/Paths.cpp

```diff
@@ -0,0 +1,554 @@
+                                                platformName,
+                                                archName, nullptr)) {
+    return result;
+  }
+
+  // <rootPath>/libexec/swift
+  if (const char *result = _swift_tryAuxExePath(name,
+                                                rootPath,
+                                                "libexec", "swift",
+                                                nullptr)) {
+    return result;
+  }
+
+  // <rootPath>/libexec/swift/<arch>
+  if (const char *result = _swift_tryAuxExePath(name,
+                                                rootPath,
+                                                "libexec", "swift",
+                                                archName, nullptr)) {
+    return result;
+  }
+
+  // <rootPath>/bin
+  if (const char *result = _swift_tryAuxExePath(name,
+                                                rootPath,
+                                                "bin", nullptr)) {
+    return result;
+  }
+
+  // <rootPath>/bin/<arch>
+  if (const char *result = _swift_tryAuxExePath(name,
+                                                rootPath,
+                                                "bin",
+                                                archName, nullptr)) {
+    return result;
+  }
+
+  // Otherwise, look in the root itself
+  return _swift_tryAuxExePath(name, rootPath, nullptr);
+}
+
+namespace {
+
+char *
+_swift_getAuxExePathIn(const char *path, const char *name)
+{
+#ifdef _WIN32
+  size_t nameLen = std::strlen(name);
+  char *nameWithSuffix = nullptr;
+  if (nameLen > 4 && strcmp(name + nameLen - 4, ".exe") != 0) {
+    nameWithSuffix = (char *)std::malloc(nameLen + 4 + 1);
+    std::memcpy(nameWithSuffix, name, nameLen);
+    std::memcpy(nameWithSuffix + nameLen, ".exe", 4 + 1);
+
+    name = nameWithSuffix;
+  }
+#endif
+
+  char *fullPath = _swift_joinPaths(path, name, nullptr);
+
+#ifdef _WIN32
+  if (nameWithSuffix)
+    std::free(nameWithSuffix);
+#endif
+
+  return fullPath;
+}
+
+const char *
+_swift_tryAuxExePath(const char *name, const char *path, ...)
+{
+  va_list val;
+  char *fullPath;
+  va_start(val, path);
+  fullPath = _swift_joinPathsV(path, val);
+  va_end(val);
+
+  if (_swift_exists(fullPath)) {
+    char *result = _swift_getAuxExePathIn(fullPath, name);
+
+    if (_swift_exists(result)) {
+      std::free(fullPath);
+
+      return result;
+    }
+
+    std::free(result);
+  }
+
+  std::free(fullPath);
+
+  return nullptr;
+}
+
+#if !defined(_WIN32) || defined(__CYGWIN__)
+void
+_swift_initRuntimePath(void *) {
+  const char *path;
+
+#if APPLE_OS_SYSTEM
+  path = dyld_image_path_containing_address(_swift_initRuntimePath);
+#else
+  Dl_info dli;
+  int ret = ::dladdr((void *)_swift_initRuntimePath, &dli);
+
+  if (!ret) {
+    swift::fatalError(/* flags = */ 0,
+                      "Unable to obtain Swift runtime path\n");
+  }
+
+  path = dli.dli_fname;
+#endif
+
+  runtimePath = ::strdup(path);
```

Comment:
Do we need to do any canonicalization here? I think that probably gets done for us by dyld but not sure about other platforms.

Reply:
(Saleem and I had already had a discussion about the situation on Windows, which predictably is the awkward case.)